src/net_processing.cpp

```diff
@@ -2908,14 +2908,15 @@ void ProcessMessage(
                 // We will continue to reject this tx since it has rejected
                 // parents so avoid re-requesting it from other peers.
                 recentRejects->insert(tx.GetHash());
+                recentRejects->insert(tx.GetWitnessHash());
```

Comment:
We set `fRejectedParents` based on parent's _txid_ in `recentRejects`. If parent has been previously announced and rejected based on wtxid only we won't found it. So we won't add this orphan and may fetch it again. I don't see this case covered in new test ?

Reply:
Also if parent txA is rejected due to a mutated witness, child txB with a _valid_ witness will be pinned in `recentRejects`. If parent txA is re-announced and accepted with a new witness, propagation of child txB will still fail until next block, assuming there is a unique valid witness known. So in multi-party, competing scenario, you introduce a _slight_ delay vector. I think that's okay if you can repeat this game at each block.

Reply:
> We set `fRejectedParents` based on parent's _txid_ in `recentRejects`. If parent has been previously announced and rejected based on wtxid only we won't found it. So we won't add this orphan and may fetch it again. I don't see this case covered in new test ?

Yes, this is an unfortunate problem, but I don't see a way to fix the redownloading of transactions with failed parent without implementing a new p2p protocol to support efficiently communicating the unconfirmed wtxid's of a given transaction (in order to know whether to reprocess a given transaction).

Note that before this PR, we have this same problem not just for child transactions of a reject parent, but for _all_ transactions that have witnesses which fail our policy checks.

Reply:
In theory we could distinguish whether failure is due to a definitely-not-witness-related reason, and in that case, add both the txid and wtxid to the recentRejects filter. I don't think this accomplishes much, because it won't occur in non-adverserial scenarios, and in adverserial ones, the attacker can just use invalid witnesses if he wants to cause harm with invalid transactions.

Reply:
I think we could try to special case the common scenario that we expect to happen at some point: some transaction spending a segwit version 1 output is rejected by (older) software as nonstandard, and then child transactions are relayed and redownloaded from each upgraded peer that announces it because the parent transaction's txid wouldn't be found in the reject filter.  We could improve this behavior by treating failures in `AreInputsStandard()` as indicative of the txid being definitely bad, and therefore add the txid to the reject filter (even though the transaction has a witness).

I think my preference would be to defer this kind of somewhat-tricky-to-reason-about change to a future PR, since this PR seems ready to me and is already an improvement, but if others feel differently we should discuss.

Reply:
This issue is only a bandwidth waste and this PR is already a huge improvement on this side, that's all good to defer this to future work, orphan parent handling is already an area known to be improved.